vllm/v1/spec_decode/ngram_proposer_gpu.py

```diff
@@ -0,0 +1,465 @@
+        )
+
+        # Apply combined mask to results. Expand mask explicitly to avoid
+        # relying on broadcasting behavior that can confuse torch.compile.
+        mask = combined_mask.unsqueeze(1).expand(-1, self.k)
+        draft_tokens = torch.where(mask, results, draft_tokens)
+
+        is_empty_draft_tokens = (draft_tokens == 0).all(dim=1)
+
+        return draft_tokens, is_empty_draft_tokens
+
+    def load_model(self, *args, **kwargs):
+        """No model to load for N-gram proposer."""
+        pass
+
+
+class NgramProposerGPU:
+    def __init__(self, vllm_config: VllmConfig, device: torch.device, runner=None):
+        assert vllm_config.speculative_config is not None
+        assert vllm_config.speculative_config.prompt_lookup_min is not None
+        assert vllm_config.speculative_config.prompt_lookup_max is not None
+
+        compilation_config = CompilationConfig(
+            level=3,
+            custom_ops=["none"],
+            splitting_ops=[],
+            compile_sizes=[],
+            inductor_compile_config={
+                "enable_auto_functionalized_v2": False,
+                "max_autotune": True,
+                "aggressive_fusion": True,
+                "triton.autotune_pointwise": True,
+                "coordinate_descent_tuning": True,
+                "use_mixed_mm": False,
+            },
+            use_cudagraph=False,
+        )
+
+        self.vllm_config = VllmConfig(compilation_config=compilation_config)
+
+        self.min_n = vllm_config.speculative_config.prompt_lookup_min
+        self.max_n = vllm_config.speculative_config.prompt_lookup_max
+        self.k = vllm_config.speculative_config.num_speculative_tokens
+        self.max_model_len = vllm_config.model_config.max_model_len
+        self.max_num_seqs = vllm_config.scheduler_config.max_num_seqs
+        self.vocab_size = vllm_config.model_config.get_vocab_size()
+        self.device = device
+
+        self.kernel = NgramGPUKernel(
+            vllm_config=vllm_config, prefix="ngram_gpu_kernel", device=device
+        )
+        self.device = device
+        self.kernel.to(device)
+        self.kernel.eval()
+
+        self._dummy_run()
+
+    def _dummy_run(self):
+        token_ids, num_tokens, sampled_flags, valid_mask = self._generate_dummy_data(
+            batch_size=self.max_num_seqs,
+            max_seq_len=self.max_model_len,
+            vocab_size=self.vocab_size,
+            pattern_len=self.k,
+            repetition_rate=0.5,
+            device=self.device,
+        )
+
+        combined_mask = sampled_flags & valid_mask & (num_tokens >= self.min_n)
+
+        for _ in range(3):
+            with set_forward_context(None, self.vllm_config):
+                _, _ = self.kernel(num_tokens, token_ids, combined_mask)
+
+    def _generate_dummy_data(
+        self,
+        batch_size: int,
+        max_seq_len: int,
+        vocab_size: int = 152064,
+        pattern_len: int = 3,
+        repetition_rate: float = 0.5,
```

Comment:
```suggestion
        vocab_size: int,
        pattern_len: int,
        repetition_rate: float,
```
Don't provide defaults for these variables, they should always be provided based on configs. If a default is needed, the config can default to a certain value.